handlers/product-switch-api/src/frequencySwitchEndpoint.ts

```diff
@@ -0,0 +1,745 @@
+): Promise<FrequencySwitchPreviewResponse> {
+	const { ratePlan, charge } = candidateCharge;
+	const result = await processFrequencySwitch(
+		zuoraClient,
+		subscription,
+		ratePlan,
+		charge,
+		productCatalog,
+		targetBillingPeriod,
+		true,
+		today,
+	);
+	return frequencySwitchPreviewResponseSchema.parse(result);
+}
+
+/**
+ * Execute a frequency switch (non-preview). Mirrors processFrequencySwitch with preview=false.
+ */
+export async function executeFrequencySwitch(
+	zuoraClient: ZuoraClient,
+	subscription: ZuoraSubscription,
+	candidateCharge: { ratePlan: RatePlan; charge: RatePlanCharge },
+	productCatalog: ProductCatalog,
+	targetBillingPeriod: 'Month' | 'Annual',
+	today: dayjs.Dayjs,
+): Promise<FrequencySwitchResponse> {
+	const { ratePlan, charge } = candidateCharge;
+	const result = await processFrequencySwitch(
+		zuoraClient,
+		subscription,
+		ratePlan,
+		charge,
+		productCatalog,
+		targetBillingPeriod,
+		false,
+		today,
+	);
+	return frequencySwitchResponseSchema.parse(result);
+}
+
+/**
+ * Get the appropriate product rate plan Id for the target billing period
+ * Uses the productRatePlanId to look up the product in the catalog (more reliable than productName)
+ * and retrieves the product rate plan Id for the target billing period.
+ *
+ * @param productCatalog Product catalog to look up rate plans
+ * @param currentRatePlan Current rate plan from the subscription (contains productRatePlanId)
+ * @param targetBillingPeriod Target billing period ('Month' or 'Annual')
+ * @returns Product rate plan Id for the target billing period
+ * @throws Error if product details cannot be found or target rate plan doesn't exist
+ */
+function getTargetRatePlanId(
+	productCatalog: ProductCatalog,
+	currentRatePlan: RatePlan,
+	targetBillingPeriod: 'Month' | 'Annual',
+): string {
+	const productCatalogHelper = new ProductCatalogHelper(productCatalog);
+
+	logger.log(
+		`Finding target rate plan for productRatePlanId: ${currentRatePlan.productRatePlanId}`,
+	);
+
+	const productDetails = productCatalogHelper.findProductDetails(
+		currentRatePlan.productRatePlanId,
+	);
+
+	if (!productDetails) {
+		throw new Error(
+			`Product rate plan ID '${currentRatePlan.productRatePlanId}' not found in product catalog`,
+		);
+	}
+
+	const targetRatePlanKey = getCatalogRatePlanName(targetBillingPeriod);
+	logger.log(
+		`Determined target rate plan key '${targetRatePlanKey}' for requested billing period '${targetBillingPeriod}'`,
+	);
+
+	// Use getAllProductDetails to find the target rate plan with proper type checking
+	const targetRatePlanDetail = productCatalogHelper
+		.getAllProductDetails()
+		.find(
+			(detail) =>
+				detail.zuoraProduct === productDetails.zuoraProduct &&
+				detail.productRatePlan === targetRatePlanKey,
+		);
+
+	if (!targetRatePlanDetail) {
+		throw new Error(
+			`Rate plan ${targetRatePlanKey} not found for product ${productDetails.zuoraProduct}`,
+		);
+	}
+
+	return targetRatePlanDetail.id;
+}
```

Comment:
I'd rather have an allow list of switches we support, rather than trying to search for rate plans based on their name in the catalog.  Something like this I think would work:

```suggestion
type TargetBillingPeriod = 'Month' | 'Annual';
function getTargetRatePlanId(
	productCatalog: ProductCatalog,
	currentRatePlan: RatePlan,
	targetBillingPeriod: TargetBillingPeriod,
): string {
	const validSwitches: Record<string, Partial<Record<TargetBillingPeriod, string>>> = {
		[productCatalog.SupporterPlus.ratePlans.Monthly.id]: { Annual: productCatalog.SupporterPlus.ratePlans.Annual.id },
	};
	const validSwitchFrom = getIfDefined(validSwitches[currentRatePlan.productRatePlanId], `Product rate plan ID '${currentRatePlan.productRatePlanId}' does not have a valid switch`);
	return getIfDefined(validSwitchFrom[targetBillingPeriod], `missing valid switch for requested billing period '${targetBillingPeriod}'`);
}
```